extensions-contrib/opentelemetry-extensions/src/main/java/org/apache/druid/data/input/opentelemetry/protobuf/OpenTelemetryMetricsProtobufReader.java

```diff
@@ -0,0 +1,205 @@
+import io.opentelemetry.proto.metrics.v1.NumberDataPoint;
+import org.apache.druid.data.input.InputEntityReader;
+import org.apache.druid.data.input.InputRow;
+import org.apache.druid.data.input.InputRowListPlusRawValues;
+import org.apache.druid.data.input.MapBasedInputRow;
+import org.apache.druid.data.input.impl.ByteEntity;
+import org.apache.druid.data.input.impl.DimensionsSpec;
+import org.apache.druid.java.util.common.CloseableIterators;
+import org.apache.druid.java.util.common.parsers.CloseableIterator;
+import org.apache.druid.java.util.common.parsers.ParseException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+public class OpenTelemetryMetricsProtobufReader implements InputEntityReader
+{
+
+  private final ByteEntity source;
+  private final String metricDimension;
+  private final String valueDimension;
+  private final String metricAttributePrefix;
+  private final String resourceAttributePrefix;
+  private final DimensionsSpec dimensionsSpec;
+
+  public OpenTelemetryMetricsProtobufReader(
+      DimensionsSpec dimensionsSpec,
+      ByteEntity source,
+      String metricDimension,
+      String valueDimension,
+      String metricAttributePrefix,
+      String resourceAttributePrefix
+  )
+  {
+    this.dimensionsSpec = dimensionsSpec;
+    this.source = source;
+    this.metricDimension = metricDimension;
+    this.valueDimension = valueDimension;
+    this.metricAttributePrefix = metricAttributePrefix;
+    this.resourceAttributePrefix = resourceAttributePrefix;
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read()
+  {
+    return CloseableIterators.withEmptyBaggage(readAsList().iterator());
+  }
+
+  List<InputRow> readAsList()
+  {
+    try {
+      return parseMetricsData(MetricsData.parseFrom(source.getBuffer()));
+    }
+    catch (InvalidProtocolBufferException e) {
+      throw new ParseException(e, "Protobuf message could not be parsed");
+    }
+  }
+
+  private List<InputRow> parseMetricsData(final MetricsData metricsData)
+  {
+    return metricsData.getResourceMetricsList()
+        .stream()
+        .flatMap(resourceMetrics -> {
+          Map<String, Object> resourceAttributes = resourceMetrics.getResource()
+              .getAttributesList()
+              .stream()
+              .collect(Collectors.toMap(kv -> resourceAttributePrefix + kv.getKey(), kv -> kv.getValue().getStringValue()));
+          return resourceMetrics.getInstrumentationLibraryMetricsList()
+              .stream()
+              .flatMap(libraryMetrics -> libraryMetrics.getMetricsList()
+                  .stream()
+                  .flatMap(metric -> parseMetric(metric, resourceAttributes).stream()));
+        })
+        .collect(Collectors.toList());
+  }
+
+  private List<InputRow> parseMetric(Metric metric, Map<String, Object> resourceAttributes)
+  {
+    List<InputRow> inputRows;
```

Comment:
can we mark this `final` to guarantee that there is only a single code path that initializes this variable?